src/details/ArborX_Ray.hpp

```diff
@@ -289,6 +296,184 @@ KOKKOS_INLINE_FUNCTION float overlapDistance(Ray const &ray,
+    // We need to rotate the triangle and the ray to be in the z = constant
+    // plane First compute the normal to the triangle: cross product of ab and
+    // ac
+    auto normal = crossProduct(ab, ac);
+    Ray::normalize(normal);
+    // Check if the rotation is necessary
+    auto normal_x = normal[0];
+    auto normal_y = normal[1];
+    auto normal_z = normal[2];
+    Point rotated_triangle_a;
+    Point rotated_triangle_b;
+    Point rotated_triangle_c;
+    Ray rotated_ray;
+    if ((normal_x > -epsilon && normal_x < epsilon) &&
+        (normal_y > -epsilon && normal_y < epsilon) &&
+        ((normal_z > 1. - epsilon && normal_z < 1. + epsilon) ||
+         (normal_z > -1. - epsilon && normal_z < -1. + epsilon)))
+    {
+      rotated_triangle_a = triangle.a;
+      rotated_triangle_b = triangle.b;
+      rotated_triangle_c = triangle.c;
+      rotated_ray = ray;
+    }
+    else
+    {
+      float angle = std::acos((double)normal[2]);
+      auto axis = crossProduct(normal, {0., 0., 1.});
+      Ray::normalize(axis);
+      rotated_triangle_a = rotatePoint(triangle.a, rotation_quat);
+      rotated_triangle_b = rotatePoint(triangle.b, rotation_quat);
+      rotated_triangle_c = rotatePoint(triangle.c, rotation_quat);
+      rotated_ray.origin() = rotatePoint(ray.origin(), rotation_quat);
+      auto rotated_direction = rotatePoint(
+          {ray.direction()[0], ray.direction()[1], ray.direction()[2]},
+          rotation_quat);
+      rotated_ray.direction() = {rotated_direction[0], rotated_direction[1],
+                                 rotated_direction[2]};
+    }
+
+    // Check that they are in the same plane
+    if (rotated_triangle_a[2] - rotated_ray.origin()[2] > -epsilon &&
+        rotated_triangle_a[2] - rotated_ray.origin()[2] < epsilon)
+    {
+      // Intersection with ab
+      if (rayEdgeIntersect(rotated_ray, rotated_triangle_a, rotated_triangle_b,
+                           t))
+        return true;
+      // Intersection with ac
+      if (rayEdgeIntersect(rotated_ray, rotated_triangle_a, rotated_triangle_c,
+                           t))
+        return true;
+      // Intersection with bc
+      if (rayEdgeIntersect(rotated_ray, rotated_triangle_b, rotated_triangle_c,
+                           t))
+        return true;
+    }
+
+    return false;
+  }
+
+  auto const inv_det = 1 / det;
+
+  auto const s = makeVector(triangle.a, ray.origin());
+  u = inv_det * dotProduct(s, p);
+  if (u < 0.f || u > 1.f)
+  {
+    return false;
+  }
+
+  auto const q = crossProduct(s, ab);
+  v = inv_det * dotProduct(ray.direction(), q);
+  if (v < 0.f || u + v > 1.f)
+  {
+    return false;
+  }
+
+  t = inv_det * dotProduct(ac, q);
+
+  if (t < epsilon)
+    return false;
+
+  return true;
+}
+
+KOKKOS_INLINE_FUNCTION bool intersects(Ray const &ray, Triangle const &triangle)
+{
+  float t;
+  float u;
+  float v;
+  return rayTriangleIntersect(ray, triangle, t, u, v);
```

Comment:
Should we return `rayTriangleIntersect(ray, triangle, t, u, v) && (t >= 0)` for consistency with the `intersects` overload we already have?

Reply:
This is already embedded in `rayTriangleIntersect`, i.e., it is `false` if `t<0`.  

One thing is that I kept the interface for `rayTriangleIntersect` but if the ray intersects the edge `u` and `v` are not set. We may want to remove `u` and `v` from the interface because it's enough to have `t` to find the point of intersections.